File: src/voidx/tools/registry.py
"""Tool registry — every tool typed, all dispatch quantified."""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict

from voidx.tools.base import ToolContext, ToolResult  # noqa: F401 — re-export
from voidx.tools.file import FileReadTool, ManageTool, WriteTool, FileReplaceTool
from voidx.tools.git import GitTool
from voidx.tools.lsp import LspFormatTool, LspTool
from voidx.tools.search import FindTool, SearchTool
from voidx.tools.bash import BashTool
from voidx.tools.task_status import TaskStatusTool
from voidx.tools.todo import TodoWriteTool
from voidx.tools.skills import SkillsTool
from voidx.tools.web import WebFetchTool, WebSearchTool
from voidx.tools.clarify import ClarifyTool
from voidx.tools.checkpoint import PlanCheckpointTool
from voidx.tools.workflow import WorkflowTool
from voidx.tools.compact import CompactContextTool
from voidx.tools.document import DocumentTool


class ToolDef(BaseModel):
    """A registered tool definition — everything the LLM needs to call it."""
    model_config = ConfigDict(arbitrary_types_allowed=True)

    id: str
    description: str
    parameters: dict  # JSON Schema for the LLM


class ToolRegistry:
    """Manages all available tools. No dynamic discovery — everything explicit."""

    _HIDDEN_FROM_LLM = frozenset({"git"})

    def __init__(self, settings=None, tracker=None) -> None:
        self._tools: dict[str, ToolDef] = {}
        self._instances: dict[str, object] = {}
        self._settings = settings
        self._tracker = tracker
        self._register_builtins()

    def _register_builtins(self) -> None:
        for cls in [
            FileReadTool, ManageTool, WriteTool, FileReplaceTool,
            GitTool,
            FindTool, SearchTool,
            LspTool, LspFormatTool,
            ClarifyTool, PlanCheckpointTool, WorkflowTool, CompactContextTool, DocumentTool,
        ]:
            instance = cls()
            self.register(instance.id, instance, instance.description, instance.parameters_schema())
        # Shell tool — platform-exclusive registration
        import os as _os
        if _os.name == "nt":
            from voidx.tools.powershell import PowerShellTool
            shell_cls = PowerShellTool
        else:
            shell_cls = BashTool
        shell_instance = shell_cls()
        self.register(shell_instance.id, shell_instance, shell_instance.description, shell_instance.parameters_schema())
        # Tools with optional dependency injection
        todo_tool = TodoWriteTool(tracker=self._tracker)
        self.register(todo_tool.id, todo_tool, todo_tool.description, todo_tool.parameters_schema())
        task_status_tool = TaskStatusTool(tracker=self._tracker)
        self.register(task_status_tool.id, task_status_tool, task_status_tool.description, task_status_tool.parameters_schema())
        skills_tool = SkillsTool(settings=self._settings)
        self.register(skills_tool.id, skills_tool, skills_tool.description, skills_tool.parameters_schema())
        retry_config = self._settings.get_retry_config() if self._settings else None
        wf = WebFetchTool(settings=self._settings, retry_config=retry_config)
        self.register(wf.id, wf, wf.description, wf.parameters_schema())
        ws = WebSearchTool(settings=self._settings)
        self.register(ws.id, ws, ws.description, ws.parameters_schema())

    def register(self, tool_id: str, instance: object, description: str, parameters: dict) -> None:
        """Register a tool dynamically (e.g. agent tool injected at runtime)."""
        self._tools[tool_id] = ToolDef(
            id=tool_id, description=description, parameters=parameters,
        )
        self._instances[tool_id] = instance

    def list(self) -> list[ToolDef]:
        return list(self._tools.values())

    def get(self, tool_id: str):
        return self._instances.get(tool_id)

    def get_def(self, tool_id: str) -> ToolDef | None:
        return self._tools.get(tool_id)

    def unregister_prefix(self, prefix: str) -> None:
        for tool_id in [tid for tid in self._tools if tid.startswith(prefix)]:
            self._tools.pop(tool_id, None)
            self._instances.pop(tool_id, None)

    def filter_tools(self, allowed_ids: set[str] | list[str] | tuple[str, ...]) -> None:
        """Retain only the tools listed in allowed_ids."""
        allowed = set(allowed_ids)
        for tool_id in [tid for tid in self._tools if tid not in allowed]:
            self._tools.pop(tool_id, None)
            self._instances.pop(tool_id, None)

    def filtered_copy(self, allowed_ids: set[str] | list[str] | tuple[str, ...]) -> "ToolRegistry":
        """Return a registry view containing existing tool defs and instances."""
        allowed = set(allowed_ids)
        clone = ToolRegistry(settings=self._settings, tracker=self._tracker)
        clone._tools = {
            tool_id: tool_def
            for tool_id, tool_def in self._tools.items()
            if tool_id in allowed
        }
        clone._instances = {
            tool_id: instance
            for tool_id, instance in self._instances.items()
            if tool_id in allowed
        }
        return clone

    def loop_filtered_copy(self, *, workflow_enabled: bool = False) -> "ToolRegistry":
        """Return the closed-world tool view for automatic runtime-backed /loop wakeups."""
        allowed = {
            "read",
            "find",
            "search",
            "lsp",
            "document",
            "websearch",
            "webfetch",
            "mcp",
            "skill",
            "bash",
        }
        if workflow_enabled:
            allowed.update({"workflow", "task_status", "todo"})
        return self.filtered_copy(allowed)

    def ids(self) -> list[str]:
        return list(self._tools.keys())

    def tools_for_llm(self) -> list[dict]:
        """Generate OpenAI/Anthropic-compatible tool definitions."""
        result = []
        for t in self._tools.values():
            if t.id in self._HIDDEN_FROM_LLM:
                continue
            # The MCP gateway passes through third-party arguments, and legacy
            # wrappers may carry non-strict external schemas. Keep MCP non-strict.
            is_mcp = t.id == "mcp" or t.id.startswith("mcp__")
            result.append({
                "type": "function",
                "function": {
                    "name": t.id,
                    "description": t.description,
                    "parameters": t.parameters,
                    **({"strict": True} if not is_mcp else {}),
                },
            })
        return result

    async def execute_tool(self, tool_id: str, args: dict, ctx: ToolContext) -> ToolResult:
        """Execute a tool by id. Returns typed result."""
        tool = self.get(tool_id)
        if not tool:
            return ToolResult(output=f"Unknown tool: {tool_id}. Available: {self.ids()}")
        return await tool.execute(args, ctx)

File: crates/voidx-tools/src/registry.rs
//! Tool registry — every tool typed, all dispatch quantified.
//!
//! Ported from `src/voidx/tools/registry.py`.

use crate::base::{Tool, ToolContext, ToolResult};
use crate::error::ToolError;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolDef {
    pub id: String,
    pub description: String,
    pub parameters: serde_json::Value,
}

/// Manages all available tools.
pub struct ToolRegistry {
    tools: HashMap<String, Arc<dyn Tool>>,
    defs: HashMap<String, ToolDef>,
}

impl ToolRegistry {
    pub fn new() -> Self {
        let mut reg = Self {
            tools: HashMap::new(),
            defs: HashMap::new(),
        };
        reg.register_builtins();
        reg
    }

    fn register_builtins(&mut self) {
        // Register all built-in tool implementations
        self.register(Arc::new(crate::bash::BashTool));
        self.register(Arc::new(crate::file_ops::FileReadTool));
        self.register(Arc::new(crate::file_ops::FileWriteTool));
        self.register(Arc::new(crate::file_ops::FileEditTool));
        self.register(Arc::new(crate::grep::GrepTool));
        self.register(Arc::new(crate::glob::GlobTool));
        self.register(Arc::new(crate::webfetch::WebFetchTool::default()));
        self.register(Arc::new(crate::websearch::WebSearchTool::default()));
        // RepoMap
        self.register(Arc::new(crate::repomap::RepoMapTool));
        // Todo & Task Status
        self.register(Arc::new(crate::todo::TodoWriteTool));
        self.register(Arc::new(crate::task_status::TaskStatusTool));
        // LSP tools
        self.register(Arc::new(crate::lsp::LspDiagnosticsTool));
        self.register(Arc::new(crate::lsp::LspSymbolsTool));
        self.register(Arc::new(crate::lsp::LspDefinitionTool));
        self.register(Arc::new(crate::lsp::LspReferencesTool));
        self.register(Arc::new(crate::lsp::LspFormatTool));
    }

    /// Register a tool dynamically.
    pub fn register(&mut self, tool: Arc<dyn Tool>) {
        let id = tool.id().to_string();
        let def = ToolDef {
            id: id.clone(),
            description: tool.description().to_string(),
            parameters: tool.parameters_schema(),
        };
        self.defs.insert(id.clone(), def);
        self.tools.insert(id, tool);
    }

    pub fn register_external(&mut self, tool: Arc<dyn Tool>) {
        self.register(tool);
    }

    /// Remove tools whose id starts with the given prefix.
    pub fn unregister_prefix(&mut self, prefix: &str) {
        let ids: Vec<String> = self
            .defs
            .keys()
            .filter(|id| id.starts_with(prefix))
            .cloned()
            .collect();
        for id in ids {
            self.defs.remove(&id);
            self.tools.remove(&id);
        }
    }

    /// List all tool definitions for LLM consumption.
    pub fn list(&self) -> Vec<&ToolDef> {
        self.defs.values().collect()
    }

    /// Generate OpenAI/Anthropic-compatible tool definitions.
    pub fn tools_for_llm(&self) -> Vec<serde_json::Value> {
        self.defs
            .values()
            .map(|t| {
                serde_json::json!({
                    "type": "function",
                    "function": {
                        "name": t.id,
                        "description": t.description,
                        "parameters": t.parameters,
                        "strict": true,
                    }
                })
            })
            .collect()
    }

    /// Look up a tool definition by id.
    pub fn get_def(&self, tool_id: &str) -> Option<&ToolDef> {
        self.defs.get(tool_id)
    }

    /// Return all tool ids.
    pub fn ids(&self) -> Vec<&str> {
        self.tools.keys().map(|s| s.as_str()).collect()
    }

    /// Retain only the tools listed in allowed_ids.
    pub fn filter_tools(&mut self, allowed_ids: &[&str]) {
        let allowed: std::collections::HashSet<&str> = allowed_ids.iter().copied().collect();
        let to_remove: Vec<String> = self
            .defs
            .keys()
            .filter(|id| !allowed.contains(id.as_str()))
            .cloned()
            .collect();
        for id in to_remove {
            self.defs.remove(&id);
            self.tools.remove(&id);
        }
    }

    /// Check if a tool is registered.
    pub fn has(&self, tool_id: &str) -> bool {
        self.tools.contains_key(tool_id)
    }

    /// Execute a tool by id.
    pub async fn execute(
        &self,
        tool_id: &str,
        args: serde_json::Value,
        ctx: &ToolContext,
    ) -> Result<ToolResult, ToolError> {
        match self.tools.get(tool_id) {
            Some(tool) => tool.execute(args, ctx).await,
            None => Err(ToolError::NotFound(tool_id.to_string())),
        }
    }
}

impl Default for ToolRegistry {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_registry_has_builtins() {
        let reg = ToolRegistry::new();
        let ids = reg.ids();
        assert!(ids.contains(&"bash"));
        assert!(ids.contains(&"file_read"));
        assert!(ids.contains(&"file_write"));
        assert!(ids.contains(&"file_edit"));
        assert!(ids.contains(&"grep"));
        assert!(ids.contains(&"glob"));
        assert!(ids.contains(&"repo_map"));
        assert!(ids.contains(&"todo"));
        assert!(ids.contains(&"task_status"));
        assert!(ids.contains(&"lsp_diagnostics"));
        assert!(ids.contains(&"lsp_symbols"));
        assert!(ids.contains(&"lsp_definition"));
        assert!(ids.contains(&"lsp_references"));
        assert!(ids.contains(&"lsp_format"));
    }

    #[test]
    fn test_tools_for_llm_produces_valid_schema() {
        let reg = ToolRegistry::new();
        let tools = reg.tools_for_llm();
        let bash = tools
            .iter()
            .find(|t| t["function"]["name"] == "bash")
            .unwrap();
        assert_eq!(bash["type"], "function");
        assert_eq!(bash["function"]["strict"], true);
        assert!(bash["function"]["parameters"]["properties"]["command"].is_object());
    }
}
